x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TokenBackwardsCompatibilityIT.java

```diff
@@ -14,197 +14,365 @@
-        // clusters
-        Request createTokenRequest = new Request("POST", "/_security/oauth2/token");
-        createTokenRequest.setJsonEntity(
-                "{\n" +
-                "    \"username\": \"test_user\",\n" +
-                "    \"password\": \"x-pack-test-password\",\n" +
-                "    \"grant_type\": \"password\"\n" +
-                "}");
-        Response response = client().performRequest(createTokenRequest);
-        assertOK(response);
-        Map<String, Object> responseMap = entityAsMap(response);
-        String token = (String) responseMap.get("access_token");
-        assertNotNull(token);
-        assertTokenWorks(token);
+        // Creates two access and refresh tokens and stores them in the token_backwards_compatibility_it index to be used for tests in the
+        // mixed/upgraded clusters
+        Map<String, Object> responseMap = createTokens(client(), "test_user", "x-pack-test-password");
+        String accessToken = (String) responseMap.get("access_token");
+        assertNotNull(accessToken);
+        assertAccessTokenWorks(accessToken);
+        String refreshToken = (String) responseMap.get("refresh_token");
+        assertNotNull(refreshToken);
 
-        Request indexRequest1 = new Request("PUT", "token_backwards_compatibility_it/_doc/old_cluster_token1");
-        indexRequest1.setJsonEntity(
-                "{\n" +
-                "    \"token\": \"" + token + "\"\n" +
-                "}");
-        Response indexResponse1 = client().performRequest(indexRequest1);
-        assertOK(indexResponse1);
-        Request createSecondTokenRequest = new Request("POST", "/_security/oauth2/token");
-        createSecondTokenRequest.setEntity(createTokenRequest.getEntity());
-        response = client().performRequest(createSecondTokenRequest);
-        responseMap = entityAsMap(response);
-        token = (String) responseMap.get("access_token");
-        assertNotNull(token);
-        assertTokenWorks(token);
-        Request indexRequest2 = new Request("PUT", "token_backwards_compatibility_it/_doc/old_cluster_token2");
-        indexRequest2.setJsonEntity(
-                "{\n" +
-                "    \"token\": \"" + token + "\"\n" +
-                "}");
-        Response indexResponse2 = client().performRequest(indexRequest2);
-        assertOK(indexResponse2);
+        storeTokens(client(), 1, accessToken, refreshToken);
+
+        responseMap = createTokens(client(), "test_user", "x-pack-test-password");
+        accessToken = (String) responseMap.get("access_token");
+        assertNotNull(accessToken);
+        assertAccessTokenWorks(accessToken);
+        refreshToken = (String) responseMap.get("refresh_token");
+        assertNotNull(refreshToken);
+
+        storeTokens(client(), 2, accessToken, refreshToken);
+    }
+
+    public void testRefreshingTokensInOldCluster() throws Exception {
```

Comment:
Why do we test refreshing a token in the `TokenBackwardsCompatibilityIT` when all nodes are on the same version ? I'm not saying testing this is not worth it, but if this means we don't test it enough where it should, maybe we could move this test there instead?

Reply:
Tests here (`TokenBackwardsCompatibilityIT`) are odd because they have side effects (they store tokens for reuse in another test phase).
This one in particular, stores the `access` and `refresh` tokens that resulted from a refresh in the old cluster, as opposed to generating them by the password grant. The `access` token is checked that it works in the _mixed_  and _upgraded_ phases, and the `refresh` token is used in the _upgraded_ cluster.

Reply:
Aha, maybe a small comment on the fact that we want to ensure that refreshed tokens keep on working in mixed cluster will be helpful